MCE IRL implementation

qxcv opened the pull request:
Is this worth fixing + merging? I assume some things (e.g the fact that I have separate abstractions for reward models) will have to be fixed before merging, but maybe it's easier just to use your implementation.

AdamGleave commented:
For reference, here's my implementation: https://github.com/HumanCompatibleAI/population-irl/blob/master/pirl/irl/tabular_maxent.py

Main issues with my implementation are:
  - State-only. Good to have: state-action, support for features.
  - Interface for Mdp's is somewhat ad-hoc. 
  - Has a Torch dependency. Probably better to use TensorFlow for consistency with rest of codebase. Although since it's such a simple algorithm maybe don't need anything as fancy as that even (and can just use autograd).

AdamGleave commented on imitation/tests/test_tabular.py:
```diff
@@ -0,0 +1,261 @@
+    # initial reward weights
+    rmodel = LinearRewardModel(mdp.obs_dim, seed=13)
+    opt = AMSGrad(rmodel, alpha_sched=1e-2)
+    final_weights, final_counts = maxent_irl(
```
I like the test case up until here (cleverly constructed MDP). Can we do anything to sanity check `maxent_irl`?
Also inclined to split these last few lines into a separate test case since they don't seem to depend on the previous part. And make it a parameterized test, so we test different reward models and optimizers. We should ideally test all combinations for a simple "smoke test" (i.e. does it run), maybe only test combinations we expect to get good results more extensively.

AdamGleave commented on imitation/tests/test_tabular.py:
```diff
@@ -0,0 +1,261 @@
+    # amsgrad typically requires (and can deal with) a higher step size
+    agd_rmodel = LinearRewardModel(10, seed=42)
+    agd = AMSGrad(agd_rmodel, alpha_sched=1e-1)
+    for rmodel, optimiser in [(sgd_rmodel, sgd), (agd_rmodel, agd)]:
```
Use a parameterised test instead.

AdamGleave commented on imitation/tests/test_tabular.py:
```diff
@@ -0,0 +1,261 @@
+        val = f(x)
+        assert np.linalg.norm(grad) > 1
+        assert np.abs(val) > 1
+        print('Initial: val=%.3f, grad=%.3f' % (val, np.linalg.norm(grad)))
```
Use `tf.logging` or Python `logging`

qxcv commented on imitation/model_env.py:
```diff
@@ -0,0 +1,346 @@
+        reward = self.reward_matrix[old_state]
+        assert np.isscalar(reward), reward
+        # copy so that it can't be mutated in-place
+        obs = self.observation_matrix[next_state].copy()
```
The point is to support different featurisations. e.g RandomMDP has options for both one-hot observations (the identity matrix) and observations that are initially sampled from a Gaussian when the MDP is constructed. This is important, because if you just stick with the one-hot featurisation then there's no point in using anything other than a linear model for your reward function or policy. All the example envs still include "perfect information" in the sense that the observation vectors for each state are unique.

AdamGleave replied:
OK, this makes sense, but really needs documenting. Can you add it to the `observation_matrix` property?

qxcv commented:
Those comments look reasonable to me, & I'll get to them either today or on the weekend. I won't do state-action rewards because it's frustrating to fit those into a Gym-like API (what kind of "observation" do you use for actions?). One of us might figure out a clean way of doing that in the future, though.

shwang commented on imitation/model_env.py:
```diff
@@ -0,0 +1,346 @@
+    """ABC for tabular environments with known dynamics."""
+    def __init__(self):
+        self.cur_state = 0
+        self.actions_taken = 0
```
I've used the convention `n_actions_taken` in this codebase

shwang commented on imitation/model_env.py:
```diff
@@ -0,0 +1,346 @@
+import numpy as np
+
+
+class ModelBasedEnv(gym.Env, metaclass=abc.ABCMeta):
```
Is there a good reason why we should be using `ABCMeta` here? Otherwise I'd just inherit from `abc.ABC`.

shwang commented on imitation/model_env.py:
```diff
@@ -0,0 +1,346 @@
+        words, if `T` is our returned matrix, then `T[s,a,sprime]` is the
+        chance of transitioning into state `sprime` after taking action `a` in
+        state `s`."""
+        pass
```
Major nit: `pass` is unnecessary here if we have a docstring

shwang commented on imitation/tabular_irl.py:
```diff
@@ -0,0 +1,426 @@
+# ############################### #
+
+
+class RewardModel(metaclass=abc.ABCMeta):
```
`abc.ABC` over `ABCMeta` in python3

qxcv commented on imitation/tests/test_tabular.py:
```diff
@@ -0,0 +1,261 @@
+    assert np.all(pi >= 0)
+    # it (always?) has to take SOME actions
+    assert np.any(pi > 0)
+    # we do <= 1 instead of allclose() because pi actually doesn't have to be
```
Oop, you're right. This comment was for a previous version of the code that iteratively applied Ziebart's infinite horizon soft backup (Algorithm 9.1 in his thesis; see bottom of p.111 for exposition), which sometimes doesn't converge to a properly normalised distribution because it incorporates a "termination reward" that allows (and potentially encourages) the agent to terminate early. The current implementation here definitely doesn't do that.

qxcv commented on imitation/tests/test_tabular.py:
```diff
@@ -0,0 +1,261 @@
+    assert np.sum(D) <= mdp.horizon + 1e-5
+    # heuristic to make sure we're *roughly* having 90% of the state encounters
+    # we expect we should
+    assert np.sum(D) >= mdp.horizon * 0.9
```
As above, the loose checks shouldn't be there anymore.

qxcv commented on imitation/model_env.py:
```diff
@@ -0,0 +1,346 @@
+        pass
+
+
+def make_random_trans_mat(
```
Do you mean an `examples` subdir at the root, or an `examples` package in `imitation`? I'd like to have them somewhere accessible so that we have simple, pre-tested environments to run MCE IRL on. At the moment one of those environments is unit tested, and both appear in the demo notebook for MCE IRL.

AdamGleave replied:
Given that you're using them in the notebook, an `examples` package in `imitation` would make more sense. I think since `RandomMDP` is used in test cases probably OK to leave it in `model_env`.

What I want to avoid is there being a large amount of code that is potentially useful (so shouldn't just be deleted) but is not essential to the operation of the algorithms and is not tested / maintained.

qxcv commented on imitation/tabular_irl.py:
```diff
@@ -0,0 +1,426 @@
+
+
+def mce_partition_fh(env, *, R=None):
+    """Calculate V^soft, Q^soft, and pi using recurrences (9.1), (9.2), and
```
Ah, that was for an old implementation using infinite horizon recursion. Looks like you've removed the comment.

AdamGleave commented on imitation/model_env.py:
```diff
@@ -0,0 +1,345 @@
+
+    @property
+    def observation_space(self):
+        if not self._observation_space:
```
For better or worse, it is standard Gym interface to *not* have `seed()` seed the observation and action spaces.

The benefit of having a property is that the `observation_space` is not mutable by external users.

I don't see why we have the `if not self._observation_space` logic though: why not just always set it in the constructor?

qxcv replied:
I did that because `.observation_space` requires `.obs_dim` to construct, which in turn requires `.observation_matrix` (in the default implementation, at least). Setting up `.observation_matrix` is the job of the subclass, though, so if the subclass calls `super().__init__()` before doing anything else, then it won't be possible to set up `.observation_space` in the `ModelBasedEnv` constructor.

There may be a less hacky way to do this. I'll talk with Steven.

qxcv replied:
Edit: talked to Steven & decided it's not worth changing. Here's a new comment explaining why I'm doing that:

```
        # Constructing action & observation spaces requires self.n_actions and
        # self.obs_dim, which are set in subclasses. If we constructed
        # observation & action in this __init__ method, then subclasses would
        # have to call super().__init__() last to give it access to
        # obs_dim/n_actions. By constructing these lazily, we ensure that
        # subclasses can call super().__init__() at any point & still have it
        # succeed.
```

AdamGleave commented on imitation/model_env.py:
```diff
@@ -0,0 +1,345 @@
+        self.cur_state = next_state
+        self.n_actions_taken += 1
+        done = self.n_actions_taken >= self.horizon
+        reward = self.reward_matrix[old_state]
```
Given the reward is currently state-only, this does feel odd to me. However, it would make complete sense for a state-action reward, since we do not know the action until this point. So I think it's probably best to keep it as-is so we have an upgrade path to state-action or state-action-successor state rewards in the future.

AdamGleave commented on imitation/tests/test_tabular.py:
```diff
@@ -0,0 +1,252 @@
+        assert len(set(map(str, trajectories))) == 1
+
+
+def test_policy_om_random_mdp():
```
`om` stands for occupancy measure (would be helped by a docstring)

AdamGleave commented:
> Those comments look reasonable to me, & I'll get to them either today or on the weekend. I won't do state-action rewards because it's frustrating to fit those into a Gym-like API (what kind of "observation" do you use for actions?). One of us might figure out a clean way of doing that in the future, though.

I'm not sure I understand the problem here. For `step`, where rewards are returned, it seems simple to make the rewards state-action since `step` receives the `action`. The trickier part is updating the IRL algorithm. But I think you can use a state-action reward matrix in the soft Bellman backup fine; in fact, looks like we already reshape `R` to `broad_R` that is effectively state-action, so I think might be as simple as just not doing that reshaping.

qxcv commented:
> I'm not sure I understand the problem here.

The problem if that your reward function needs to use a different observation model to that exposed by the Gym env for approximate policies. e.g you'd need to add a `.reward_observation_space` attribute (for action + reward) or `.action_observation_space` (for just action representation, which could get concatenated to state representation), etc. None of that fits inside the existing Gym API. Then again, for these tabular models we have to work outside the Gym API anyway to get model info, so perhaps it doesn't matter.

AdamGleave commented:
> The problem if that your reward function needs to use a different observation model to that exposed by the Gym env for approximate policies. e.g you'd need to add a `.reward_observation_space` attribute (for action + reward) or `.action_observation_space` (for just action representation, which could get concatenated to state representation), etc. None of that fits inside the existing Gym API. Then again, for these tabular models we have to work outside the Gym API anyway to get model info, so perhaps it doesn't matter.

Ah I see, so the issue is that we might want to index `reward_matrix[o, a]` by observations `o` that are not from `env.observation_space` and actions `a` that are not from `env.action_space`?

I think we have the same issue with `env.transition_matrix`, since that's `[s, a, s']` indexed. Currently I think `mce_partition_fh` assumes `s` and `s'` are integers between `0` and `env.n_states`, and `a` between `0` and `env.n_actions`. Think it should be fine to do the same for reward for now.

qxcv commented:
No, that's not quite what I'm getting at. Originally I wanted to support arbitrary combinations of action & state features for the "observation space" of the reward model. For example, you could have a different, randomly-generated n-dimensional vector for each possible combination of state & action (like `RandomMDP` does for states at the moment), or an image-based observation that combines the action & reward together (e.g by including a depiction of the current action's effect in the image, like an arrow showing the direction the agent has chosen to move). Just as a state has many possible featurisations, each combination of state & action could also have many featurisations, and I wanted to support more-or-less arbitrary featurisations.

AdamGleave commented:
> No, that's not quite what I'm getting at. Originally I wanted to support arbitrary combinations of action & state features for the "observation space" of the reward model. For example, you could have a different, randomly-generated n-dimensional vector for each possible combination of state & action (like `RandomMDP` does for states at the moment), or an image-based observation that combines the action & reward together (e.g by including a depiction of the current action's effect in the image, like an arrow showing the direction the agent has chosen to move). Just as a state has many possible featurisations, each combination of state & action could also have many featurisations, and I wanted to support more-or-less arbitrary featurisations.

Me & Sam had a call and are now largely on the same page. My understanding is:
  - The observation and action space exposed by an environment is a modeling choice. For example, in MuJoCo, do you provide an image of the pixel rendering or the raw positions as an observation? Both are of the same underlying environment, but raw positions are much easier to learn with.
  - Because of its importance, for reproducibility it belongs in the environment, and not as e.g. a choice in the policy or reward model.
  - It's common for people to do further transformations in their policy or reward model (e.g. downsampling and conversion to grayscale in Atari). These should be declared, but are not typically a big deal. In particular, most simple transformations will just be losing information. Very complex transformations (e.g. training a computer vision model on Atari, then feeding those features into your model) are more dubious.
  - Policies receive an observation of the state and output an action. Reward models receive an observation of the state *and* action, and output a reward. 

  In general, there is no reason that the observations received by the policies and reward models need to be the same. However, the simplest option is to assume they are the same, and make the reward model's input a concatenation of the observation of the state (received by the policy) and the observation of the action. In this case many environments will work out the box, although the representation of the action may be non-ideal (often actions are discrete, and Gym's `Discrete' represents as integer, may be better to be one-hot coded).

  I was leaning towards going with the above option. Sam I think feels it's better to stick with state-only for now and decide later, possibly going for a more general option.

@qxcv let me know if I've misrepresented anything!

qxcv commented:
That's a good summary, thanks Adam!

qxcv commented on imitation/model_env.py:
```diff
@@ -0,0 +1,345 @@
+
+    def seed(self, seed=None):
+        if seed is None:
+            seed = np.random.randint(0, 1 << 31)
```
For some reason the Gym spec wants a list of seeds to be returned, so I do it that way :man_shrugging:

qxcv commented on imitation/model_env.py:
```diff
@@ -0,0 +1,345 @@
+        if seed is None:
+            seed = np.random.randint(0, 1 << 31)
+        self.rand_state = np.random.RandomState(seed=seed)
+        return [seed]
```
Gym API spec wants this

qxcv commented on imitation/model_env.py:
```diff
@@ -0,0 +1,345 @@
+        current state, current action, and next state (in that order). In other
+        words, if `T` is our returned matrix, then `T[s,a,sprime]` is the
+        chance of transitioning into state `sprime` after taking action `a` in
+        state `s`."""
```
You can add that later if you want, but remember that `.transition_matrix` may not exist when `ModelBasedEnv.__init__()` is called by the subclass.

qxcv commented on imitation/tabular_irl.py:
```diff
@@ -0,0 +1,426 @@
+        rew_params = optimiser.current_params
+        rmodel.set_params(rew_params)
+        t += 1
+        if occupancy_error_dest is not None:
```
I don't use those args in the rest of the code anyway, so I just removed them.

qxcv commented on imitation/tabular_irl.py:
```diff
@@ -0,0 +1,426 @@
+    raise TypeError("No idea how to make schedule out of '%s'" % lr_or_schedule)
+
+
+class Optimiser(metaclass=abc.ABCMeta):
```
I can use implementations from Jax, I just didn't do that the first time around because I wanted to be 100% sure I knew what the optimisers were doing. I'll go fix that now.

qxcv replied:
Also, the Jax optimiser library is a thing of beauty: https://github.com/google/jax/blob/97e16655155cc6e176e154ff932d7b2133e8cf1d/jax/experimental/optimizers.py#L180-L369

Compare that to the source of `tf.train` :stuck_out_tongue_winking_eye:

qxcv commented:
I've fixed most of the issues. Things I haven't done:

- State-action rewards. Someone else can look at that if they think they've got a good API (ideally one we can use for AIRL etc., too).
- Removing Jax. IMO it doesn't make sense to do this until stable baselines moves to TF 2.x, at which point we'll have to rewrite a lot of things anyway (& migrating from Jax to TF 2.x is probably easier than migrating from TF 1.x to TF 2.x). I replaced the custom optimisers with default Jax ones, so there's not going to be a complexity benefit to moving to TF at the moment either.

AdamGleave commented:
Thanks for fixing the outstanding issues. I agree it's worth waiting until we start migrating to TF 2.x to port from Jax.

I think the state-action reward is likely worth adding soon, but I don't think it should block this PR.

I've opened issue #14 to track enhancements such as these we may want to make in the future.

AdamGleave commented:
@shwang any additional thoughts on these changes?
